tests/functional_tests/ffront_tests/test_domain_decution.py

```diff
@@ -0,0 +1,33 @@
+import pytest
+
+from functional.common import Dimension, GridType, GTTypeError
+from functional.ffront.decorator import Program
+from functional.ffront.fbuiltins import FieldOffset
+
+
+Dim = Dimension("Dim")
+LocalDim = Dimension("LocalDim", local=True)
+
+CartesianOffset = FieldOffset("CartesianOffset", source=Dim, target=(Dim,))
+UnstructuredOffset = FieldOffset("UnstructuredOffset", source=Dim, target=(Dim, LocalDim))
+
+
+def test_domain_deduction():
```

Comment:
Might be useful to separate these more, also to document more clearly that this operates in several modes (pure deduction, overriding deduction with user choice, failing if user choice is not appropriate).